### PR TITLE
fix:add missing fields

### PR DESCRIPTION
**Pull request by trestinlsd**
## Description

----

avs module add missing fields

Fix vulnerability GO-2024-2694 in github.com/cosmos/ibc-go/v7

- Upgraded github.com/cosmos/ibc-go/v7 to version v7.4.0 to address GO-2024-2694 vulnerability.
- Fixed potential reentrancy issue in ibc-hooks.
- Updated dependencies to ensure code integrity and security.

### CODE REVIEW

**Comment by cloud8little** [comment]
Test passed with https://github.com/ExocoreNetwork/exocore/pull/64/commits/745b2783f4389cfb4cbfce5f6b0e2cc00979401f
avs owner address is added. 

```shell
exocored q avs AVSInfo exo1l7m80efn6lfkdcng47rwy2rma6xsqywu6cwe2v --home ~/.tmp-exocored
info:
  asset_id:
  - "123"
  avs_address: exo1l7m80efn6lfkdcng47rwy2rma6xsqywu6cwe2v
  avs_owner_address: exo1nkh04qzxed98megzz53xx9a4uju599zg3rw0mj
  name: avs01
  operator_address:
  - exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph
```

**Review by cloud8little — APPROVED** [review]

**Review by mikebraver — APPROVED** [review]

**Review by TimmyExogenous — COMMENTED** [review]
It seems that the register and deregister functions of the AVS can still be called by any address, right?   
If the owner address of the AVS is a parameter passed in during registration, then deregistration or updates would require signature verification of the owner address. Implementing such permission verification is relatively complex. Moreover, this PR only adds an owner address parameter initialization and does not address the permission control issues related to AVS operations. The key should be to use the "from" address in the msg call as the owner address. You can refer to the implementation of other similar functionalities for details.

**Comment by cloud8little** [comment]
Move to further milestone as the owner address needs to be well-designed.

File: precompiles/avs/avs.go [hunk 1]
```diff
@@ -84,8 +84,12 @@ func (p Precompile) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) (bz [
 	// It avoids panics and returns the out of gas error so the EVM can continue gracefully.
 	defer cmn.HandleGasError(ctx, contract, initialGas, &err)()
 
-	if method.Name == MethodAVSAction {
+	switch method.Name {
+	case MethodAVSAction:
 		bz, err = p.AVSInfoRegisterOrDeregister(ctx, evm.Origin, contract, stateDB, method, args)
+	case MethodRegisterOperatorToAVS:
```
Comment (mikebraver) [inline precompiles/avs/avs.go]: So won't you support operator deregister from AVS?

File: precompiles/avs/tx.go [hunk 2]
```diff
@@ -1,29 +1,41 @@
...
+	cmn "github.com/evmos/evmos/v14/precompiles/common"
+	"golang.org/x/xerrors"
 )
 
 const (
 	// AVSRegister defines the ABI method name for the avs
 	// related transaction.
-	MethodAVSAction = "AVSAction"
+	MethodAVSAction             = "AVSAction"
+	MethodRegisterOperatorToAVS = "registerOperatorToAVS"
```
Comment (mikebraver) [inline precompiles/avs/tx.go]: Same here

File: precompiles/avs/tx.go [hunk 3]
```diff
@@ -33,3 +45,35 @@ func (p Precompile) AVSInfoRegisterOrDeregister(
 	}
 	return method.Outputs.Pack(true)
 }
+
+func (p Precompile) OperatorInfoRegister(
```
Comment (mikebraver) [inline precompiles/avs/tx.go]: The name of this function is confused, the operator is registered via the operator module, what you'd do is to register operator to avs, shouldn't this added in operator module?

File: x/avs/keeper/operator.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	operator := &types.OperatorInfo{
+		Name:         name,
+		PublicKey:    pub,
+		IsRegistered: true,
+	}
+	store.Set(opAccAddr, k.cdc.MustMarshal(operator))
+	return nil
+}
+
+func (k *Keeper) GetOperator(ctx sdk.Context, addr string) (operator types.OperatorInfo, err error) {
```
Comment (mikebraver) [inline x/avs/keeper/operator.go]: Why not get operator in operator module?

**Review by mikebraver — CHANGES_REQUESTED** [review]
Many places need to be updated!

File: precompiles/avs/tx.go (hunk 3 above)
Comment (mikebraver) [reply precompiles/avs/tx.go]: Also need to support operator deregister from avs

File: proto/exocore/avs/tx.proto [hunk 5]
```diff
@@ -11,14 +14,62 @@ message AVSInfo {
   string name = 1;
   // address of avs
   string avs_address = 2;
-  // registered operator of avs
-  repeated string operator_address = 3;
+  //  slash address of avs
+  string slash_addr = 3;
   // the owner who has permission for avs
   string avs_owner_address = 4;
```
Comment (mikebraver) [inline proto/exocore/avs/tx.proto]: Won't there's any case that avs has multiple owners or operators?

File: precompiles/avs/types.go [hunk 6]
```diff
@@ -14,52 +14,56 @@ func (p Precompile) GetAVSParamsFromInputs(_ sdk.Context, args []interface{}) (*
...
 
-	assetID, ok := args[5].(string)
-	if !ok || assetID == "" {
-		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "uint64", action)
+	unbondingEpochs, ok := args[6].(uint64)
+	if !ok || (action != avstypes.RegisterAction && action != avstypes.DeRegisterAction) {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 6, "uint64", unbondingEpochs)
 	}
-	avsParams.AssetID = assetID
+	avsParams.UnbondingEpochs = unbondingEpochs
```
Comment (mikebraver) [inline precompiles/avs/types.go]: You defined both `unbondingEpochs` and `Epochs` in protobuf struct, why there's only unbondingEpochs set here?

File: proto/exocore/avs/tx.proto [hunk 7]
```diff
@@ -11,14 +14,62 @@ message AVSInfo {
...
+  [
+    (cosmos_proto.scalar) = "cosmos.Int",
+    (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Int",
+    (gogoproto.nullable) = false
+  ];
+  //avs epoch ,Subsequently  will be handled by the epochs module.
+  EpochInfo avs_epoch = 8;
+}
+//  The information of OperatorInfo in opt-in avs
+message OperatorInfo {
```
Comment (mikebraver) [inline proto/exocore/avs/tx.proto]: Shouldn't it store the avs meta info within this struct?

File: x/avs/keeper/avs.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	errorsmod "cosmossdk.io/errors"
+	sdkmath "cosmossdk.io/math"
+	"fmt"
+	"github.com/ExocoreNetwork/exocore/x/avs/types"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func (k *Keeper) GetAVSSupportedAssets(ctx sdk.Context, avsAddr string) ([]string, error) {
```
Comment (mikebraver) [inline x/avs/keeper/avs.go:12]: Please reuse the function `GetAVSInfo` to get avs info, and return res.AssetId then

File: x/avs/keeper/avs.go [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		return nil, errorsmod.Wrap(types.ErrNoKeyInTheStore, fmt.Sprintf("GetAVSSupportedAssets: key is %s", avsAddr))
+	}
+
+	value := store.Get([]byte(avsAddr))
+	ret := types.AVSInfo{}
+	k.cdc.MustUnmarshal(value, &ret)
+
+	return ret.AssetId, nil
+}
+func (k *Keeper) GetAVSSlashContract(ctx sdk.Context, avsAddr string) (string, error) {
```
Comment (mikebraver) [inline x/avs/keeper/avs.go:21]: Same here

**Comment by mikebraver** [comment]
Please also fix all action errors

File: x/avs/types/types.go [hunk 10]
```diff
@@ -1 +1,19 @@
 package types
+
+func ContainsString(slice []string, target string) bool {
```
Comment (mikebraver) [inline x/avs/types/types.go]: No need to re-implement it, just use `import slices` and `slices.Contains()`

File: precompiles/avs/avs.go (hunk 1 above)
Comment (trestinlsd) [reply precompiles/avs/avs.go]: Supported

File: precompiles/avs/tx.go (hunk 3 above)
Comment (trestinlsd) [reply precompiles/avs/tx.go]: fixed

File: precompiles/avs/tx.go (hunk 2 above)
Comment (trestinlsd) [reply precompiles/avs/tx.go]: fixed

File: x/avs/keeper/operator.go (hunk 4 above)
Comment (trestinlsd) [reply x/avs/keeper/operator.go]: fixed

File: proto/exocore/avs/tx.proto (hunk 5 above)
Comment (trestinlsd) [reply proto/exocore/avs/tx.proto]: supported

**Review by mikebraver — APPROVED** [review]
Can merge it after all comments and action error are resolved

File: x/avs/keeper/avs.go (hunk 8 above)
Comment (trestinlsd) [reply x/avs/keeper/avs.go:12]: fixed

File: x/avs/keeper/avs.go (hunk 9 above)
Comment (trestinlsd) [reply x/avs/keeper/avs.go:21]: fixed

File: x/avs/types/types.go (hunk 10 above)
Comment (trestinlsd) [reply x/avs/types/types.go]: fixed

File: precompiles/avs/abi.json [hunk 11]
```diff
@@ -1,35 +1,40 @@
...
+				"internalType": "uint64",
+				"name": "minimumDelegation",
+				"type": "uint64"
 			},
 			{
-				"internalType": "string",
-				"name": "assetID",
-				"type": "string"
+				"internalType": "uint64",
+				"name": "unbondingEpochs",
```
Comment (bwhour) [inline precompiles/avs/abi.json]: unbondingEpoch:  it's better named it **unbondingTime** or **unbondingPeriod**

File: precompiles/avs/avs.sol [hunk 12]
```diff
@@ -14,11 +14,20 @@ IAVSManager constant AVSMANAGER_CONTRACT = IAVSManager(
 /// @custom:address 0x0000000000000000000000000000000000000902
 interface IAVSManager {
     function AVSAction(
```
Comment (bwhour) [inline precompiles/avs/avs.sol:16]: Why not name it RegisterAVS or InitializeAVS? AVSAction doesn't clearly indicate what the interface is doing.

File: precompiles/avs/avs.sol [hunk 13]
```diff
@@ -14,11 +14,20 @@ IAVSManager constant AVSMANAGER_CONTRACT = IAVSManager(
...
+        string[] memory avsOwnerAddress,
         string memory avsName,
-        string memory avsAddress,
-        string memory operatorAddress,
+        string memory slashContractAddr,
+        string[] memory assetID,
         uint64 action,
-        string memory avsOwnerAddress,
-        string memory assetID
+        uint64 minimumDelegation,
```
Comment (bwhour) [inline precompiles/avs/avs.sol]: is this minimumSelfDelegation ?

File: precompiles/avs/avs.sol [hunk 14]
```diff
@@ -14,11 +14,20 @@ IAVSManager constant AVSMANAGER_CONTRACT = IAVSManager(
...
+        uint64 unbondingEpochs
```
Comment (bwhour) [inline precompiles/avs/avs.sol]: unbondingTime/unbondingPeriod would be better.

File: precompiles/avs/tx.go [hunk 15]
```diff
@@ -1,29 +1,56 @@
...
 
 import (
+	"fmt"
+	"slices"
+
+	errorsmod "cosmossdk.io/errors"
+
+	exocmn "github.com/ExocoreNetwork/exocore/precompiles/common"
+	util "github.com/ExocoreNetwork/exocore/utils"
+	avstypes "github.com/ExocoreNetwork/exocore/x/avs/keeper"
```
Comment (bwhour) [inline precompiles/avs/tx.go]: this should be **avskeeper** not avstypes. please change the related places.

File: precompiles/avs/tx.go [hunk 16]
```diff
@@ -1,29 +1,56 @@
...
 	_ vm.StateDB,
 	method *abi.Method,
 	args []interface{},
 ) ([]byte, error) {
 	// parse the avs input params first.
 	avsParams, err := p.GetAVSParamsFromInputs(ctx, args)
+	if err != nil {
+		return nil, errorsmod.Wrap(err, "parse args error")
+	}
+	avsAddress, err := util.ProcessAvsAddress(contract.Address().String())
```
Comment (bwhour) [inline precompiles/avs/tx.go]: ProcessAVSAddress

File: precompiles/avs/tx.go [hunk 17]
```diff
@@ -1,29 +1,56 @@
...
 const (
-	// AVSRegister defines the ABI method name for the avs
-	// related transaction.
-	MethodAVSAction = "AVSAction"
+	MethodAVSAction      = "AVSAction"
+	MethodOperatorAction = "OperatorOptAction"
 )
 
 // AVSInfoRegister register the avs related information and change the state in avs keeper module.
 func (p Precompile) AVSInfoRegisterOrDeregister(
```
Comment (bwhour) [inline precompiles/avs/tx.go]: **AVSInfoRegisterOrDeregister:**
Although this function name describes its functionality, it is a bit long and not specific enough. Consider using a more concise and clear name.It is recommended to change it to RegisterOrDeregisterAVSInfo, which better follows the action + object naming convention for functions.

**OperatorBindingAvs:**
This function name can also be more specific, clearly indicating that it binds an operator to AVS.
It is recommended to change it to BindOperatorToAVS, which more clearly expresses the function's purpose.

File: x/avs/keeper/avs.go [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+		return nil, errorsmod.Wrap(err, fmt.Sprintf("GetAVSSupportedAssets: key is %s", avsAddr))
+	}
+
+	return avsInfo.Info.AssetId, nil
+}
+
+func (k *Keeper) GetAVSSlashContract(ctx sdk.Context, avsAddr string) (string, error) {
+	avsInfo, err := k.GetAVSInfo(ctx, avsAddr)
+	if err != nil {
+		return "", errorsmod.Wrap(err, fmt.Sprintf("GetAVSSupportedAssets: key is %s", avsAddr))
```
Comment (bwhour) [inline x/avs/keeper/avs.go]: this function log GetAVSSupportedAssets is not right

File: x/avs/keeper/avs.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+	return avsInfo.Info.SlashAddr, nil
+}
+
+// GetAVSMinimumSelfDelegation returns the USD value of minimum self delegation, which
+// is set for operator
+func (k *Keeper) GetAVSMinimumSelfDelegation(ctx sdk.Context, avsAddr string) (sdkmath.LegacyDec, error) {
+	avsInfo, err := k.GetAVSInfo(ctx, avsAddr)
+	if err != nil {
+		return sdkmath.LegacyNewDec(0), errorsmod.Wrap(err, fmt.Sprintf("GetAVSSupportedAssets: key is %s", avsAddr))
```
Comment (bwhour) [inline x/avs/keeper/avs.go]: here function log : GetAVSSupportedAssets

File: x/avs/keeper/keeper.go [hunk 20]
```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
+		return errorsmod.Wrap(types.ErrAlreadyRegistered, fmt.Sprintf("the error input arg is:%s", params.AvsAddress))
+	}
+
+	avs := &types.AVSInfo{
+		Name:               params.AvsName,
+		AvsAddress:         params.AvsAddress,
+		SlashAddr:          params.SlashContractAddr,
+		AvsOwnerAddress:    params.AvsOwnerAddress,
+		AssetId:            params.AssetID,
+		AvsUnbondingEpochs: uint32(params.MinimumDelegation),
```
Comment (bwhour) [inline x/avs/keeper/keeper.go]: this param has a  lot of types in different places: unit32, uint64, int64. please have a check.
this assignment is strange : 

AvsUnbondingEpochs: uint32(params.MinimumDelegation),
MinimumDelegation:  sdk.NewIntFromUint64(params.UnbondingEpochs),

File: x/avs/keeper/keeper.go [hunk 21]
```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 }
 
 func (k Keeper) AVSInfoUpdate(ctx sdk.Context, params *AVSRegisterOrDeregisterParams) error {
+	avsInfo, _ := k.GetAVSInfo(ctx, params.AvsAddress)
+
+	action := params.Action
+
+	if action == RegisterAction && avsInfo != nil {
+		return errorsmod.Wrap(types.ErrAlreadyRegistered, fmt.Sprintf("the error input arg is:%s", params.AvsAddress))
```
Comment (bwhour) [inline x/avs/keeper/keeper.go]: can replace this with specific error  " the avsaddress is :%s",

File: x/avs/keeper/keeper.go [hunk 22]
```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
+		MinimumDelegation:  sdk.NewIntFromUint64(params.UnbondingEpochs),
+		AvsEpoch:           nil,
+		OperatorAddress:    nil,
+	}
+	if action == RegisterAction && avsInfo == nil {
+		return k.SetAVSInfo(ctx, avs)
+	}
+
+	if avsInfo == nil {
+		return errorsmod.Wrap(types.ErrUnregisterNonExistent, fmt.Sprintf("the error input arg is:%s", avsInfo))
```
Comment (bwhour) [inline x/avs/keeper/keeper.go]: avsInfo == nil so this line has no meaning print:   fmt.Sprintf("the error input arg is:%s", avsInfo)

File: x/avs/keeper/keeper.go [hunk 23]
```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
-
-	action := params.Action
-
-	if action == RegisterAction {
-		return k.SetAVSInfo(ctx, params.AvsName, params.AvsAddress, operatorAddress, params.AssetID)
+		return errorsmod.Wrap(delegationtypes.ErrOperatorNotExist, fmt.Sprintf("AVSInfoUpdate: invalid operator address:%s", operatorAddress))
 	}
 	avsInfo, err := k.GetAVSInfo(ctx, params.AvsAddress)
-	if err != nil {
+	if err != nil || avsInfo.GetInfo() == nil {
```
Comment (bwhour) [inline x/avs/keeper/keeper.go]: there are two lines call **avsInfo.GetInfo()**, can we call once ?

File: x/avs/keeper/keeper.go [hunk 24]
```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
 		return errorsmod.Wrap(err, fmt.Sprintf("error occurred when get avs info %s", avsInfo))
 	}
-	if avsInfo.Info.AvsOwnerAddress != params.AvsOwnerAddress {
-		return errorsmod.Wrap(err, fmt.Sprintf("not qualified to deregister %s", avsInfo))
+	avs := avsInfo.GetInfo()
```
Comment (bwhour) [inline x/avs/keeper/keeper.go:97]: see above comment

File: x/avs/keeper/keeper.go [hunk 25]
```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
-		store.Set(avsAddr, bz)
-		return nil
-	}
-	value := store.Get(avsAddr)
-	ret := &types.AVSInfo{}
-	k.cdc.MustUnmarshal(value, ret)
-	ret.OperatorAddress = append(ret.OperatorAddress, operatorAddress)
-	ret.AssetId = append(ret.AssetId, assetID)
+
+	AVSInfo := avs
```
Comment (bwhour) [inline x/avs/keeper/keeper.go]: this AVSInfo  variable is the same as types.AVSInfo

**Review by bwhour — CHANGES_REQUESTED** [review]
some comments left need to be fixed.

File: precompiles/avs/abi.json (hunk 11 above)
Comment (trestinlsd) [reply precompiles/avs/abi.json]: fixed

File: precompiles/avs/avs.sol (hunk 13 above)
Comment (trestinlsd) [reply precompiles/avs/avs.sol]: Minimum stake quantity for Operator

File: precompiles/avs/avs.sol (hunk 14 above)
Comment (trestinlsd) [reply precompiles/avs/avs.sol]: fixed

File: precompiles/avs/tx.go (hunk 15 above)
Comment (trestinlsd) [reply precompiles/avs/tx.go]: fixed

File: precompiles/avs/tx.go (hunk 16 above)
Comment (trestinlsd) [reply precompiles/avs/tx.go]: fixed

File: precompiles/avs/tx.go (hunk 17 above)
Comment (trestinlsd) [reply precompiles/avs/tx.go]: fixed

File: x/avs/keeper/avs.go (hunk 18 above)
Comment (trestinlsd) [reply x/avs/keeper/avs.go]: fixed

File: x/avs/keeper/avs.go (hunk 19 above)
Comment (trestinlsd) [reply x/avs/keeper/avs.go]: fixed

File: x/avs/keeper/keeper.go (hunk 20 above)
Comment (trestinlsd) [reply x/avs/keeper/keeper.go]: fixed

File: x/avs/keeper/keeper.go (hunk 21 above)
Comment (trestinlsd) [reply x/avs/keeper/keeper.go]: fixed

File: x/avs/keeper/keeper.go (hunk 22 above)
Comment (trestinlsd) [reply x/avs/keeper/keeper.go]: unregistration must be done with data available

File: x/avs/keeper/keeper.go (hunk 23 above)
Comment (trestinlsd) [reply x/avs/keeper/keeper.go]: fixed

File: x/avs/keeper/keeper.go (hunk 24 above)
Comment (trestinlsd) [reply x/avs/keeper/keeper.go:97]: fixed

File: x/avs/keeper/keeper.go (hunk 25 above)
Comment (trestinlsd) [reply x/avs/keeper/keeper.go]: fixed

**Review by bwhour — APPROVED** [review]

**Pull request merged**